text/0000-inline-asm.md

```diff
@@ -0,0 +1,892 @@
+  - Floating-point exception flags in `FPSCR` (IDC, IXC, UFC, OFC, DZC, IOC).
+- AArch64
+  - Condition flags (`NZCV` register).
+  - Floating-point status (`FPSR` register).
+- RISC-V
+  - Floating-point exception flags in `fcsr` (`fflags`).
+
+> Note: As a general rule, these are the flags which are *not* preserved when performing a function call.
+
+## Mapping to LLVM IR
+
+The direction specification maps to a LLVM constraint specification as follows (using a `reg` operand as an example):
+
+* `in(reg)` => `r`
+* `out(reg)` => `=&r` (Rust's outputs are early-clobber outputs in LLVM/GCC terminology)
+* `inout(reg)` => `=&r,0` (an early-clobber output with an input tied to it, `0` here is a placeholder for the position of the output)
+* `lateout(reg)` => `=r` (Rust's late outputs are regular outputs in LLVM/GCC terminology)
+* `inlateout(reg)` => `=r, 0` (cf. `inout` and `lateout`)
+
+If an `inout` is used where the output type is smaller than the input type then some special handling is needed to avoid LLVM issues. See [this bug][issue-65452].
+
+As written this RFC requires architectures to map from Rust constraint specifications to LLVM constraint codes. This is in part for better readability on Rust's side and in part for independence of the backend:
+
+* Register classes are mapped to the appropriate constraint code as per the table above.
+* `imm` operands are formatted and injected directly into the asm string.
+* `sym` is mapped to `s` for statics and `X` for functions.
+* a register name `r1` is mapped to `{r1}`
+* additionally mappings for register classes are added as appropriate (cf. [llvm-constraint])
+* `lateout` operands with an `_` expression that are specified as an explicit register are converted to LLVM clobber constraints. For example, `lateout("r1") _` is mapped to `~{r1}` (cf. [llvm-clobber]).
+* If the `nomem` flag is not set then `~{memory}` is added to the clobber list. (Although this is currently ignored by LLVM)
+* If the `preserves_flags` flag is not set then the following are added to the clobber list:
+  - (x86) `~{dirflag}~{flags}~{fpsr}`
+  - (ARM/AArch64) `~{cc}`
+
+For some operand types, we will automatically insert some modifiers into the template string.
+* For `sym` and `imm` operands, we automatically insert the `c` modifier which removes target-specific modifiers from the value (e.g. `#` on ARM).
+* On AArch64, we will warn if a value smaller than 64 bits is used without a modifier since this is likely a bug (it will produce `x*` instead of `w*`). Clang has this same warning.
+* On ARM, we will automatically add the `P` or `q` LLVM modifier for `f64`, `v64` and `v128` passed into a `vreg`. This will cause those registers to be formatted as `d*` and `q*` respectively.
+
+Additionally, the following attributes are added to the LLVM `asm` statement:
+
+* The `nounwind` attribute is always added: unwinding from an inline asm block is not allowed (and not supported by LLVM anyways).
```

Comment:
This seems like a big footgun, e.g., if an `asm!` block calls a function that unwinds.

Reply:
I would assume LLVM supports unwinding from inline assembly, otherwise why would it support using the `invoke` instruction as well as the `call` instruction to represent inline assembly in LLVM IR?

Reply:
> I would assume LLVM supports unwinding from inline assembly, otherwise why would it support using the `invoke` instruction as well as the `call` instruction to represent inline assembly in LLVM IR?

Turns out that LLVM has `invoke asm` to help support unwinding from inline assembly, but [currently assumes inline assembly doesn't unwind](https://github.com/llvm/llvm-project/blob/master/llvm/test/Transforms/Inline/inline-invoke-with-asm-call.ll) and doesn't properly translate `invoke asm` in the backends.

Reply:
> This seems like a big footgun, e.g., if an `asm!` block calls a function that unwinds the behavior is undefined. We could at least `abort` the program with a nice error message?

1) Aborting would require additional code around every asm block, for something that almost no asm block will *ever* do.

2) If LLVM doesn't support unwind from assembly, we can't do otherwise without adding such support to LLVM first.

3) We could potentially support an `unwind` attribute in the future, if LLVM adds such support.

Reply:
> Aborting would require additional code around every asm block,

Can't we just use landing pads, just like we do for `#[unwind(abort)]`? That does increase code-size, and code size can impact execution time, but beyond this particular impact, the non-exceptional path should be fast.

> We could potentially support an unwind attribute in the future, if LLVM adds such support.

This would mean that some footguns are optin (nomem, nostack, pure, readonly, ...) while others are opt-out (unwind), and that feels inconsistent.

Reply:
@gnzlbg
> > Aborting would require additional code around every asm block,
> 
> Can't we just use landing pads, just like we do for `#[unwind(abort)]`?

That would be one variant of the "additional code" I was referring to. And please no. Unwinding from assembly is incredibly rare, and the far far more common case of *not* unwinding from assembly should not incur overhead for it (whether by having to write "nounwind" on every asm block or by emitting a landing pad where we otherwise wouldn't have).

One of my use cases for inline assembly involves writing the smallest possible programs in Rust, using inline assembly to make system calls. (I hope to see crates for making syscalls become usable on stable soon after inline assembly becomes usable on stable.) Such programs would use `unwind=abort`, and should not contain any landing pads.

> > We could potentially support an unwind attribute in the future, if LLVM adds such support.
> 
> This would mean that some footguns are optin (nomem, nostack, pure, readonly, ...) while others are opt-out (unwind), and that feels inconsistent.

unwind isn't a footgun, here. Many assembly blocks will want to access memory, or the stack, or have side effects, or affect the flags; it's not at all unusual to need to do so, and having to pass a flag to allow doing so correctly is a footgun. The vast majority of inline assembly blocks don't unwind, and most don't even call a function; doing so is incredibly rare, and it doesn't seem at all unreasonable to require a flag to do so.

Reply:
> Can't we just use landing pads, just like we do for #[unwind(abort)]? That does increase code-size, and code size can impact execution time, but beyond this particular impact, the non-exceptional path should be fast.

That won't work because LLVM assumes that inline asm can't unwind. This means that it will optimize your landing pads away since they are "unreachable", or it may simply not emit the unwinding tables at all for the asm.